### PR TITLE
Plane: Fix level-off and throttle without airspeed sensor

### PR DESCRIPTION
**Pull request by menschel**
This PR does intend to fix Issues https://github.com/ArduPilot/ardupilot/issues/28685 and https://github.com/ArduPilot/ardupilot/issues/28775

This patch intends to fix two issues with auto takeoff mission item for
planes without an airspeed sensor.

Level-off fix:
- Correct the level-off condition by factor 2 because during level-off,
  the pitch and therefore the climb-rate decreses linearly.
  Mathmatically the plane therefore only climbs half of the altitude in the expected time
  and is not guaranteed to converge to target altitude by itself.
- Add flag for level-off stage and let TECS handle the final climb.

Throttle without airspeed sensor:
- Remove conditional evaluation of airspeed sensor to enable traditional
  TECS-controlled throttle range also without airspeed sensor.
  This restores the behavior of ArduPlane 4.5.7.

It is to be discussed if the flag THROTTLE_RANGE in parameter TKOFF_OPTIONS is to be reversed to keep
the behavior from Plane 4.5.7 consistent with Plane 4.6 onwards.
It could be something like FORCE_MAX_THROTTLE instead.

Latest measurements with this change cherry-picked onto ArduPilot-4.6:

[24-12-01_09-12-18.bin](https://github.com/menschel/ardupilot/blob/stash/build/bin/Arduplane/24-12-01_09-12-18.bin)
[24-12-01_09-16-52.bin](https://github.com/menschel/ardupilot/blob/stash/build/bin/Arduplane/24-12-01_09-16-52.bin)
[24-12-01_09-21-56.bin](https://github.com/menschel/ardupilot/blob/stash/build/bin/Arduplane/24-12-01_09-21-56.bin)
level_off_fix
throttle_fix

### CODE REVIEW

**Comment by menschel** [comment]
I would appreciate feedback on the final meter issue. The plane is above target altitude of 50m but exits takeoff mission item 5-6 seconds later when 51m is reached. Am I testing wrong?

**Comment by menschel** [comment]
After a couple of iterations, I came to the conclusion, that Navigation / TECS iterates much faster on the target altitude, if the target altitude of the waypoint is elevated by 1m.
The other conclusion is that the calculation of the level-off switch time is flawed.
Linearly raping out pitch, also linearly ramps out climb-rate. Therefore it cannot reach the target-altitude by math. It can only gain half of the altitude in that time. I will replace the measurements and the figures later.

**Comment by menschel** [comment]
One remark, I have absolutely no clue why with that mission, I get 4A current during takeoff. The exact same mission yields 3A in Plane 4.5.7

**Comment by menschel** [comment]
I'm satisfied with the takeoff behavior now. There is no visible difference from 4.5.7. The parameter needs to be discussed though. I argue the parameter should be "inverted", so any user migrating from 4.5.7 to 4.6 experiences the known behavior.

File: ArduPlane/takeoff.cpp [hunk 1]
```diff
@@ -309,13 +308,17 @@ int16_t Plane::get_takeoff_pitch_min_cd(void)
 
         // are we entering the region where we want to start levelling off before we reach takeoff alt?
         if (auto_state.sink_rate < -0.1f) {
-            float sec_to_target = (remaining_height_to_target_cm * 0.01f) / (-auto_state.sink_rate);
+            float sec_to_target = (remaining_height_to_target_cm * 0.01f * 2.0f) / MIN(TECS_controller.get_max_climbrate(), -auto_state.sink_rate);
```
Comment (Georacer) [inline ArduPlane/takeoff.cpp:311]: The decrease of pitch is not linear; that much we've seen from the SITL tests. Can you provide some justification for this formula?

**Review by Georacer — COMMENTED** [review]
Thank you for your effort @menschel !

As you have noticed, we chose to address the level-off bug in a much simpler way, implemented here: https://github.com/ArduPilot/ardupilot/pull/28792

Regarding the throttle behaviour, we intentionally changed it from 4.5 to 4.6 here: https://github.com/ArduPilot/ardupilot/pull/27758
In 4.5, the throttle during takeoff would be managed by TECS in TAKEOFF mode but set fixed to `TKOFF_THR_MAX` in AUTO mode takeoffs.
We opted to keep the 2nd behaviour.

If you really want the old throttle behaviour for TAKEOFF mode back, we would probably need to make use of another `TKOFF_OPTIONS` bit. That would need to be done carefully, testing all corner cases.

It's not high on my priority list. If you want to implement that, you're very welcome of course.
If you believe the 4.6 behaviour is dangerous for some reason, naturally we'd want to hear your argument.

File: ArduPlane/takeoff.cpp (hunk 1 above)
Comment (menschel) [reply ArduPlane/takeoff.cpp:311]: I make the case that the scalar creates the linear dependency between remaining height and pitch.
Integrating A*X results in (A/2)*X² and that's where the factor 2 comes from.
Setting both equal results in X=1/2 for the gained altitude.

The MIN() is a sanity check to limit climb-rate to the max value known to TECS.
It is not dependent on angle but on the power to weight ratio.

I don't know what happens in 4.5.7 but there it is not linear.

**Comment by menschel** [comment]
Ahem, you have mixed Takeoff-mode and Auto-mode takeoff mission item.
I fly Auto-mode missions to have reproducible results between tests.
In 4.5 TECS manages throttle in Auto-mode during takeoff mission item and it works great.

**Comment by menschel** [comment]
This PR has been stale for 2 weeks. I assume it is not welcome then.

**Comment by rmackay9** [comment]
Hi @menschel,

Sorry for the delay, it takes us a while to get to the bottom of issues some times though.  This item is still on our 4.6 issues list so we're going to keep looking into it a bit more.
